src/Microsoft.Extensions.Hosting/HostBuilder.cs

```diff
@@ -0,0 +1,211 @@
+            {
+                buildAction(configBuilder);
+            }
+            _hostConfiguration = configBuilder.Build();
+        }
+
+        private void CreateHostingEnvironment()
+        {
+            _hostingEnvironment = new HostingEnvironment()
+            {
+                ApplicationName = _hostConfiguration[HostDefaults.ApplicationKey],
+                EnvironmentName = _hostConfiguration[HostDefaults.EnvironmentKey] ?? EnvironmentName.Production,
+                ContentRootPath = ResolveContentRootPath(_hostConfiguration[HostDefaults.ContentRootKey], AppContext.BaseDirectory),
+            };
+            _hostingEnvironment.ContentRootFileProvider = new PhysicalFileProvider(_hostingEnvironment.ContentRootPath);
+        }
+
+        private string ResolveContentRootPath(string contentRootPath, string basePath)
+        {
+            if (string.IsNullOrEmpty(contentRootPath))
+            {
+                return basePath;
+            }
+            if (Path.IsPathRooted(contentRootPath))
+            {
+                return contentRootPath;
+            }
+            return Path.Combine(Path.GetFullPath(basePath), contentRootPath);
+        }
+
+        private void CreateHostBuilderContext()
+        {
+            _hostBuilderContext = new HostBuilderContext(Properties)
+            {
+                HostingEnvironment = _hostingEnvironment,
+                Configuration = _hostConfiguration
+            };
+        }
+
+        private void BuildAppConfiguration()
+        {
+            var configBuilder = new ConfigurationBuilder();
+            configBuilder.AddConfiguration(_hostConfiguration);
+            foreach (var buildAction in _configureAppConfigActions)
+            {
+                buildAction(_hostBuilderContext, configBuilder);
+            }
+            _appConfiguration = configBuilder.Build();
+            _hostBuilderContext.Configuration = _appConfiguration;
+        }
+
+        private void CreateServiceProvider()
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton(_hostingEnvironment);
+            services.AddSingleton(_hostBuilderContext);
+            services.AddSingleton(_appConfiguration);
+            services.AddSingleton<IApplicationLifetime, ApplicationLifetime>();
+            services.AddSingleton<IHostLifetime, HostProcessLifetime>();
+            services.AddOptions();
+            services.AddLogging();
+            
+            foreach (var configureServicesAction in _configureServicesActions)
+            {
+                configureServicesAction(_hostBuilderContext, services);
+            }
+
+            // We can't invoke the factory directly because we don't know the genric type at compile time.
+            // TContainerBuilder IServiceProviderFactory<TContainerBuilder>.CreateBuilder(IServiceCollection)
+            var createBuilderMethod = _serviceProviderFactory.GetType().GetMethod("CreateBuilder");
```

Comment:
Need to handle exceptions? What if there are overloads of CreateBuilder in the type? AmbiguousMatchException? In fact, this approach looks odd overall.

Reply:
I don't think we need to get too fancy here. Providing a factory at all is less common, and it's restricted to a specific interface. We can search CreateBuilder and CreateServiceProvider overloads if people ask for it. It's pretty ugly code that I'd rather avoid if we can.

As for handling exceptions, I'm not too concerned. The inputs were restricted via strong types so we know they at least match the interface signature. Reflection has good exceptions for duplicate methods so we don't need to wrap that. The only interesting error will be if someone's generics don't match between UseServiceProviderFactory and ConfigureContainer, and the reflection exception in that case is pretty clear too.

Reply:
We should be able to do this without reflection via a few internal adapter types that have non-generic interfaces. I think that would be preferable to reflection...